driver-core/src/main/java/com/datastax/driver/core/policies/ErrorAwarePolicy.java

```diff
@@ -0,0 +1,338 @@
+ *   WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ *   See the License for the specific language governing permissions and
+ *   limitations under the License.
+ */
+package com.datastax.driver.core.policies;
+
+import com.codahale.metrics.Clock;
+import com.codahale.metrics.Meter;
+import com.datastax.driver.core.*;
+import com.datastax.driver.core.exceptions.*;
+import com.google.common.annotations.VisibleForTesting;
+import com.google.common.collect.AbstractIterator;
+import org.slf4j.Logger;
+import org.slf4j.LoggerFactory;
+
+import java.util.Collection;
+import java.util.Iterator;
+import java.util.Map;
+import java.util.concurrent.ConcurrentHashMap;
+import java.util.concurrent.ConcurrentMap;
+
+/**
+ * Chainable load balancing policy that will filter out hosts based on their errors rates.
+ * <p/>
+ * The policy bases its list of hosts on a child policy on which it will compute whether this host needs to included
+ * or excluded of the query plan.
+ * <p/>
+ * The policy uses a {@link LatencyTracker} that it will register automatically on the {@link Cluster}. This tracker
+ * will not take into account successful queries and only take note of the errors produced by queries. If the
+ * accounted number of errors for a host exceeds the threshold configured in the policy, the Host will be ignored
+ * for the query plan.
+ * <p/>
+ * The policy follows the builder pattern to be created, the {@link Builder} class can be created with
+ * {@link #builder} method.
+ */
+public class ErrorAwarePolicy implements ChainableLoadBalancingPolicy {
+
+    private static final Logger logger = LoggerFactory.getLogger(ErrorAwarePolicy.class);
+
+    private final LoadBalancingPolicy childPolicy;
+
+    private final int retryPeriodMillis;
+
+    private final Clock clock;
+
+    PerHostErrorTracker errorTracker;
+
+    private ErrorAwarePolicy(Builder builder) {
+        this.childPolicy = builder.childPolicy;
+        this.retryPeriodMillis = builder.retryPeriod;
+        this.clock = builder.clock;
+        this.errorTracker = new PerHostErrorTracker(builder.errorsThreshold, builder.errorsFilter, builder.clock);
+    }
+
+    @Override
+    public LoadBalancingPolicy getChildPolicy() {
+        return childPolicy;
+    }
+
+    @Override
+    public void init(Cluster cluster, Collection<Host> hosts) {
+        childPolicy.init(cluster, hosts);
+        cluster.register(this.errorTracker);
+    }
+
+    @Override
+    public HostDistance distance(Host host) {
+        return childPolicy.distance(host);
+    }
+
+    @Override
+    public Iterator<Host> newQueryPlan(String loggedKeyspace, Statement statement) {
+        final Iterator<Host> childQueryPlan = childPolicy.newQueryPlan(loggedKeyspace, statement);
+
+        return new AbstractIterator<Host>() {
+
+            @Override
+            protected Host computeNext() {
+                long currentTime = clock.getTime();
```

Comment:
I changed this from `System.currentTimeMillis()` -> `clock.getTime()` since we happen to have a `Clock` instance, but is shouldn't make a difference since the default `Clock` will call `System.currentTimeMillis()`